challenge_11/database.rb

```diff
@@ -0,0 +1,87 @@
+### Require dependencies
+require 'cgi'
+require 'rack/handler/puma'
+require 'rack'
+require 'sqlite3'
+
+# Struct to define what a Blog looks like
+Blog = Struct.new(:title, :content, keyword_init: true)
+
+SEED_BLOGS = [
+  Blog.new(title: 'My awesome blog!', content: 'my favourite HTML tags are <p> and <script>'),
+  Blog.new(title: 'Another cool blog!', content: 'my favourite HTML tags are <br> and <hr>')
+]
+
+# Create or open SQLite database
+database = SQLite3::Database.new('application.sqlite3', results_as_hash: true)
+
+# Check if we need to create our table
+existing_tables = database.execute('SELECT name FROM sqlite_master WHERE type="table";')
+unless existing_tables.include?({ 'name' => 'blogs'})
+  database.execute('CREATE TABLE blogs (title VARCHAR(30), content TEXT);')
+end
+
+# Alternative
+# table_exists = database.get_first_value('SELECT COUNT(*) FROM sqlite_master WHERE type="table" AND name="blogs";')
+# if table_exists == 0
+#   database.execute('CREATE TABLE blogs (title VARCHAR(30), content TEXT);')
+# end
+
+# Seed some blog data if none exists already
+if database.get_first_value('SELECT COUNT(*) FROM blogs;') == 0
+  SEED_BLOGS.each do |blog|
+    database.execute('INSERT INTO blogs (title, content) VALUES (?, ?)', [blog.title, blog.content])
+  end
+end
+
+app = lambda { |environment|
+  puts 'Rack app got a request!'
+  # Build Rack request
+  request = Rack::Request.new(environment)
+
+  # Build Rack response
+  response = Rack::Response.new
+
+  if request.get? && request.path == '/show-data'
+    response.content_type = 'text/html'
+    response.finish do
+      response.write '<ul>'
+      blogs = database.execute('SELECT * FROM blogs;').map { |data| Blog.new(data) }
+      blogs.each do |blog|
```

Comment:
A small thing vaguely related to ideas we’ve discussed before: you can pass a block to `#execute`, in which case it [calls `#each` on the underlying `ResultSet`](https://github.com/sparklemotion/sqlite3-ruby/blob/v1.4.2/lib/sqlite3/database.rb#L198-L200) and yields each row to the block. This is interesting because `ResultSet#each` just [repeatedly calls `#next`](https://github.com/sparklemotion/sqlite3-ruby/blob/v1.4.2/lib/sqlite3/resultset.rb#L133) which itself [calls `#step`](https://github.com/sparklemotion/sqlite3-ruby/blob/v1.4.2/lib/sqlite3/resultset.rb#L108) on the statement object, which then [ultimately calls `sqlite3_step`](https://github.com/sparklemotion/sqlite3-ruby/blob/v1.4.2/ext/sqlite3/statement.c#L127) from the [SQLite C interface](https://www.sqlite.org/c3ref/step.html).

As you can see from the SQLite documentation, this is a sort of incremental API that is designed to be called repeatedly as the results are processed. I don’t know anything about the implementation details here, but it’s reasonable to assume that SQLite can break up the work of a large query and perform it in small chunks in response to repeated calls of `sqlite3_step`.

So my _point_ is that, when you call `#execute` with a block, the rows are kind of “streamed” back from the SQLite library as they’re generated, which would dovetail nicely with the [streaming responses](https://github.com/adrianna-chang-shopify/learning-project/commit/f5c3d6a2b5c93a3729e0effe8142f113db6cb28b) we’ve experimented with previously. It’s not hard to imagine connecting these features together to make a snappy-feeling web application that’s able to begin serving an expensive response immediately and stream the full body back to the client while it’s still being generated from the database. Neat.

Reply:
Oh, this is really cool! Yeah, so I'd imagine that similar to what we were doing with the [streaming responses in challenge 10](https://github.com/adrianna-chang-shopify/learning-project/commit/f5c3d6a2b5c93a3729e0effe8142f113db6cb28b), we could call `#execute` with a block that would write the row to the response object, and if our response object was using `#finish` with the block form as well, this would result in that data getting written straight to the socket! I think it would look something like this?
```ruby
response.finish do
  response.write '<ul>'
  blogs = database.execute('SELECT * FROM blogs;') do |row|
    blog = Blog.new(row)
    response.write '<li>'
    response.write "<strong>Title: #{CGI.escape_html(blog.title)}</strong>, Content: #{CGI.escape_html(blog.content)}"
    response.write '</li>'
  end
  response.write '</ul>'
end
```

So if we had a million blogs in our db, this would be a really neat way to provide a responsive UX.